### PR TITLE
add data provider design doc proposal

### PR DESCRIPTION
**Pull request by helinwang**
Fixes https://github.com/PaddlePaddle/Paddle/issues/1344

### CODE REVIEW

File: doc/design/reader/README.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
+# Python Data Provider Design Doc
```
Comment (helinwang) [inline doc/design/reader/README.md]: 这个PR里用了provider, generator, reader这几个名字。太混乱。
准备改成一致用reader。

File: doc/design/reader/README.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        yield buf
+        buf = [reader.next() for x in xrange(minibatch_size)]
+
+def train(provider, mapping, batch_size, total_pass):
+	for pass_idx in range(total_pass):
+		for mini_batch in make_minibatch_generator(provider(pass_idx)): # this loop will never end in online learning.
+			do_forward_backward(mini_batch, mapping)
+```
+
+This is just an example implementation, more complicated logic like data prefetch could be implemented.
```
Comment (helinwang) [inline doc/design/reader/README.md]: TODO: 加上combine data reader的例子，比如GAN需要combine true image data reader。

Comment (helinwang) [inline doc/design/reader/README.md]: prefetch可以通过`make_prefetch_reader`来做，`make_prefetch_reader`可以放在`paddle.data` package下面。

File: doc/design/reader/README.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+# ...
+
+paddle.train(paddle.data.mnist, ["image", "label"], 128, 10, ...)
+```
+## Q & A
+
+### Why return only a single entry, but not a mini batch?
+
+If return a mini batch, data provider need to take care of batch size. But batch size is a concept for training, it makes more sense for user to specify batch size as a parameter for `train`.
```
Comment (reyoung) [inline doc/design/reader/README.md]: 这里有一个问题，就是用户有可能已经就把数据存储成一个batch一个文件了(因为如果一个数据一个文件，文件太细碎，但是所有数据一个文件，文件又太大)。。这样，用户的reader，为什么不能返回一个batch size的数据呢？

```python
def process(filename):
    data_batch = pickle.load(open(filename))
    yield data_batch
```

相反，我觉得我们的接口应该是一个返回iterator的函数，这个iterator必须返回一个batch的数据。而对于只返回一条数据的情况，只是batch_size为1的batch。

这样做可以有更大的灵活性。特别是，这样我们中间所有的higher-order functions的接口都是一样的了(返回值都是一个batch)。

File: doc/design/reader/README.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+Concretely, always return a single entry make reusing existing data providers much easier (e.g., if existing data provider return not a single entry but 3 entries, training code will be more complex because it need to handle cases like batch size 2).
```
Comment (reyoung) [inline doc/design/reader/README.md]: `batch_size`是train函数的参数么？我之前的理解里，data这个接口一次返回多少数据，batch size就是多少。这样做的好处是

* batch_size不只会在train函数里面，比如批量的model inference之类的东西，也都要求设置batch size。如果batch_size放到train函数的参数里面，那其他需要batch size的函数，是不是也要设置batch size呢？
* 其实，我们只需要写一个高阶函数, resize_batch_size，就可以处理用户的Data函数里面的batch_size，和实际需要的batch size不一致的情况。

这种高阶函数组合可以非常简单的完成非常复杂的工作。譬如，我们有两个函数 shuffle, resize_batch_size

```python
def shuffle(func):
    pass

def resize_batch_size(func, batch_size):
    pass
```

* shuffle会对iterator里面的每一个元素做shuffle，也就是shuffle一个batch内部的数据
* resize_batch_size会改变batch_size的大小，如果batch_size=-1的话，就将iterator里所有数据放到在一个batch里面。

那么，带缓存的，全数据 shuffle的 mnist reader就可以是

```python
reader = resize_batch_size(
    func=shuffle(func=resize_batch_size(
                        func=mnist_reader, 
                        batch_size=-1)),
    batch_size=256)
```

而，如果内存太小，我们的数据没办法全部放到内存里，那么就改一下最内部的`resize_batch_size`即可

```python
reader = resize_batch_size(
    func=shuffle(func=resize_batch_size(
                        func=mnist_reader, 
                        batch_size=1024*1024)),  # memory limit.
    batch_size=256)
```

File: doc/design/reader/README.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+provider = lambda : image_provier("/path/to/data/")
+paddle.train(provider, ["image", "label"], ...)
+```
+
+### How is `paddle.train` implemented
+
+An example implementation of paddle.train could be:
+
+```python
+def make_minibatch_generator(reader, minibatch_size):
```
Comment (reyoung) [inline doc/design/reader/README.md]: 这种高阶函数是不是应该再trainer之外指定？即trainer里面不需要知道什么batch_size，shuffle之类的东西。对于trainer来说，reader就是很简单的调用就好了。

```python
for data_batch in reader():
    pass
```

而高阶函数的组合拼装，如果觉得放在trainer外面，用户自己写比较麻烦，那么我们也可以提供一个builder，或者简单的helper

```python
mnist_reader = ...  # read one sample generator

trainer.train(reader = buffered_batch_shuffle_reader(mnist_reader, batch_size=32), ...) 
```

File: doc/design/reader/README.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+	images = numpy.fromfile(
+		f, 'ubyte', count=n * 28 * 28).reshape((n, 28 * 28)).astype('float32')
+	images = images / 255.0 * 2.0 - 1.0
+	labels = numpy.fromfile(l, 'ubyte', count=n).astype("int")
+	for i in xrange(n):
+		yield [images[i, :], labels[i]] # a single entry of data is created each time
+	f.close()
+
+# use python lambda to change image_provier into a function with no parameters.
+provider = lambda : image_provier("/path/to/data/")
```
Comment (wangkuiyi) [inline doc/design/reader/README.md]: 这里不用lambda了吧。一个调用了yield的函数就是一个generator。一旦被调用，不是执行函数体，而是返回一个iterator。这个iterator每次被访问的时候执行generator的函数体。

File: doc/design/reader/README.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+```
+## Q & A
+
+### Why return only a single entry, but not a mini batch?
+
+If return a mini batch, data provider need to take care of batch size. But batch size is a concept for training, it makes more sense for user to specify batch size as a parameter for `train`.
+
+Concretely, always return a single entry make reusing existing data providers much easier (e.g., if existing data provider return not a single entry but 3 entries, training code will be more complex because it need to handle cases like batch size 2).
+
+### How to create custom data provider
```
Comment (wangkuiyi) [inline doc/design/reader/README.md]: 这里有多个问题：

1. how to create a data reader -- write a function which calls yield.
2. how to manipulate a data reader -- call make_concat_reader, make_buffered_reader.
3. how to compose existing data readers -- call 
    ```python
    make_composite_reader(create_text_reader("labels.txt"),
                          create_image_reader("/tmp/mnist/image_file_list.txt"))
    ```

File: doc/design/reader/README.md (hunk 2 above)
Comment (wangkuiyi) [inline doc/design/reader/README.md]: 我建议叫`make_buffered_reader(reader, buffer_size)`或者`paddle.reader.buffered(reader, buffer_size)`

Comment (wangkuiyi) [inline doc/design/reader/README.md]: combine 说不清楚是多个reader顺序被调用以贡献统一路数据，还是并行被调用以返回多路数据。

如果是前者，可以叫 `make_interleaving_reader(readers)`或者`paddle.reader.interleave(readers)`

如果是后者，可以叫`make_composite_reader(readers)`或者`paddle.reader.compose(readers)`

File: doc/design/reader/README.md (hunk 5 above)
Comment (wangkuiyi) [inline doc/design/reader/README.md]: `make_minibatch_reader`或者叫做 `paddle.reader.batched`可能也可以

File: doc/design/reader/README.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+```python
+def image_provider(path):
```
Comment (wangkuiyi) [inline doc/design/reader/README.md]: image_provider作为一个Python generator不能“provide data”。而是在被调用之后返回一个（可以被称为provider的）iterable，来provide data。所以 image_provider 应该叫做 create_image_reader。

File: doc/design/reader/README.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+	f = open(path)
+	images = numpy.fromfile(
+		f, 'ubyte', count=n * 28 * 28).reshape((n, 28 * 28)).astype('float32')
+	images = images / 255.0 * 2.0 - 1.0
+	labels = numpy.fromfile(l, 'ubyte', count=n).astype("int")
```
Comment (wangkuiyi) [inline doc/design/reader/README.md]: fromfile的参数 l 是在哪儿定义的呢？

File: doc/design/reader/README.md [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+	for i in xrange(n):
+		yield [images[i, :], labels[i]] # a single entry of data is created each time
```
Comment (wangkuiyi) [inline doc/design/reader/README.md]: 这个例子是不是应该是两个reader——一个image reader和一个text reader——分别从两个文件里读取数据，然后用 paddle.reader.compose 合并为一个reader比较好。这样可以展示readers的可复用性。

File: doc/design/reader/README.md [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+```
+
+## Usage
+
+data provider, mapping from data provider column to data layer, batch size and number of total pass will be passed into `paddle.train`:
+
+```python
+# two data layer is created:
+image_layer = paddle.layer.data("image", ...)
+label_layer = paddle.layer.data("label", ...)
```
Comment (helinwang) [inline doc/design/reader/README.md]: data layer感觉不需要同名。我好像没有想象出来什么情况需要同名。

File: doc/design/reader/README.md [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+# ...
+
+paddle.train(paddle.data.mnist, ["image", "label"], 128, 10, ...)
```
Comment (helinwang) [inline doc/design/reader/README.md]: 是的，有一个方法是
```
paddle.train({"image": image_reader, "label": label_reader}, ...)
```
之前提出来过，大家觉得image_reader和label_reader应该是强联系在一起的（image和label一一对应），分开两个reader觉得奇怪。所以就没有用这个方法。

File: doc/design/reader/README.md (hunk 6 above)
Comment (helinwang) [inline doc/design/reader/README.md]: 同意，👍

File: doc/design/reader/README.md (hunk 12 above)
Comment (helinwang) [inline doc/design/reader/README.md]: `{"image":0, "label":1}`

File: doc/design/reader/README.md (hunk 2 above)
Comment (helinwang) [inline doc/design/reader/README.md]: Need helper func `make_shuffled_reader` to shuffle some length of items: `make_shuffled_reader(reader, int)`

Comment (helinwang) [inline doc/design/reader/README.md]: TODO: add difference between `reader creator` and `reader decorator`

File: doc/design/reader/README.md (hunk 1 above)
Comment (helinwang) [inline doc/design/reader/README.md]: 决定一致叫`reader`，稍后更新PR。

File: doc/design/reader/README.md [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+## Data Provider Interface
+
+Data provider is a function with no parameter that creates a iterable (anything can be used in `for x in iterable`):
+
+```
+iterable = data_provider()
+```
+
+Element produced for the iterable should be a **single** entry of data, in format `[column_0_item, column_1_item, ...]`. Each element of the list needs to be supported data type (e.g., numpy 1d array of float32, list of int).
```
Comment (helinwang) [inline doc/design/reader/README.md]: 好的，会加上说明。（不对应network的data layer）

File: doc/design/reader/README.md (hunk 11 above)
Comment (helinwang) [inline doc/design/reader/README.md]: 你的意思是可能map到同一个column。对的。所以要用map。 ｀{"image":0, "label":1}｀，稍后PR加上。

File: doc/design/reader/README.md (hunk 3 above)
Comment (helinwang) [inline doc/design/reader/README.md]: 讨论结果每次返回一个data

File: doc/design/reader/README.md (hunk 4 above)
Comment (helinwang) [inline doc/design/reader/README.md]: 讨论结果每次返回一个data :)

File: doc/design/reader/README.md (hunk 7 above)
Comment (helinwang) [inline doc/design/reader/README.md]: Will add in PR.

File: doc/design/reader/README.md (hunk 8 above)
Comment (helinwang) [inline doc/design/reader/README.md]: Will change in PR.

File: doc/design/reader/README.md (hunk 9 above)
Comment (helinwang) [inline doc/design/reader/README.md]: A bug, will change in PR.

File: doc/design/reader/README.md (hunk 5 above)
Comment (helinwang) [inline doc/design/reader/README.md]: Yes, we need helpers.

File: doc/design/reader/README.md (hunk 2 above)
Comment (helinwang) [inline doc/design/reader/README.md]: TODO: list useful decorators, and name them according to naming convention.

Comment (helinwang) [inline doc/design/reader/README.md]: TODO: perhaps in here, or data package design doc. Add note that we should not need user to download data by a second step. E.g., `paddle.data.mnist` should download or reuse cached data.

File: doc/design/reader/README.md [hunk 14]
```diff
@@ -0,0 +1,79 @@
+# Python Data Provider Design Doc
+
+Data provider provides data for training. It will be passed into `paddle.train` as a parameter.
+
+## Data Provider Interface
+
+Data provider is a function with no parameter that creates a iterable (anything can be used in `for x in iterable`):
```
Comment (helinwang) [inline doc/design/reader/README.md]: 我觉得只要是能生成一个iterable的都可以。不必要提具体用哪个方式（generator或者是class）。

File: doc/design/reader/README.md (hunk 10 above)
Comment (helinwang) [inline doc/design/reader/README.md]: Done. Added section for combine data readers.

File: doc/design/reader/README.md (hunk 5 above)
Comment (helinwang) [inline doc/design/reader/README.md]: Maybe we don't need to expose`paddle.reader.batched`, since data reader should always return single data entry. We only need to use it internally.

File: doc/design/reader/README.md (hunk 2 above)
Comment (helinwang) [inline doc/design/reader/README.md]: It's the latter, using `paddle.reader.compose`

Comment (helinwang) [inline doc/design/reader/README.md]: All done. Have not explicitly "list useful decorators", but listed use cases and mentioned them.

File: doc/design/reader/README.md [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+
+```python
+def minibatch_decorater(reader, minibatch_size):
+	def ret():
+		r = reader()
+		buf = [r.next() for x in xrange(minibatch_size)]
+		while len(buf) > 0:
+			yield buf
+			buf = [r.next() for x in xrange(minibatch_size)]
+	return ret
```
Comment (qingqing01) [inline doc/design/reader/README.md]: 这块我测了下有点bug，  比如reader有[1,2, ...., 14] 14个数， minibatch_size=4，最后一个batch应该是[13, 14]两个数，但是这样写测试，并没有返回~

Comment (helinwang) [inline doc/design/reader/README.md]: 嗯是的，最后剩余的要不要用是一个待讨论的问题。
不保留的好处是在这个例子里，mini batch size肯定都是10（不会有2的情况）。
保留的好处是数据都会被用到。

我觉得可以train给optional选项表明要是剩余data小于batch size怎么处理。

Comment (helinwang) [inline doc/design/reader/README.md]: 决定默认不丢掉。
